### ISSUE TITLE
Create a documentation comment linter that is compatible with DartDoc

### ISSUE THREAD

**Issue by anwersayeed** [PalisadoesFoundation/talawa#1479]
We recently implemented automatic updating of the [docs.talawa.io](https://docs.talawa.io) with descriptions extracted from the comments in the Talawa code base.  This creates a wide range of HTML pages that explain the [methods, functions and classes present in each file](https://docs.talawa.io/talawa/index.html).

There are limitations with this approach.
1. It relies on the code contributors to be diligent in their comments so that they can be extracted
2. The formatting of the comments needs to be consistent.

There are many ways to approach this, but there are some general standards that are being accepted. For example there is the simple python pydocstyle linter that ensures basic compliance and the presence of comments immediately below each class, method and function declaration. This is insufficient.

Google has taken this one step further. In their python guide they explain that every method, class and function should have a [description, definitions of the arguments and definitions of the results](https://docs.talawa.io/talawa/index.html).

![image](https://user-images.githubusercontent.com/16875803/218168221-0181916c-82fa-4631-ae6b-361664477a1b.png)

The [TSDoc project](https://tsdoc.org/) attempts to do the same with `Typescript`, and have created a linter to help enforce compliance.

![image](https://user-images.githubusercontent.com/16875803/218168742-d1eea116-f393-4708-9c1c-56e6331e5b80.png)

We want to maintain the quality of our documentation by identifying a linter which has a way of creating HTML documentation.

Research a method to enforce documentation compliance for dart comments.
- Every class, method, and function must have documentation about it

Documentation should include:

- A brief description of the code
- Argument variables that are used to initialize the code with a description of what each argument does
- Variables that the code returns and a description of each variable

### Note 
1. We are currently using [DartDoc](https://pub.dev/packages/dartdoc) for generating HTML and would **GREATLY PREFER** to see whether there is any way to use it to enforce these or very similar standards
1.  For example linting to make sure that the comments comply with the [Dart Documentation Comments](https://dart.dev/guides/language/language-tour#documentation-comments) standard would be acceptable.
1. You can refer to this pull request for work that is being done by the dart team https://github.com/dart-lang/linter/pull/3820, but others may be doing something through GitHub actions or some other means

**Describe the solution you'd like**
1. Finding a linting package that can be used as part of our GitHub actions
2. Finding a way for it to be run on all files or on just one file at a time.

**Describe alternatives you've considered**
1. Writing script files. We feel that there must be an existing solution to this.

**Comment by literalEval** [comment]
@anwersayeed @palisadoes I am interested in working on this issue. I am not very proficient with linters and docs but I am willing to learn.

**Comment by palisadoes** [comment]
Are you proficient with python?

One approach would be to write a python script that reads files and makes sure there is an comment block that includes:

1. An overall description, 
1. argument descriptions and,
1. a description of the values returned,

for each method class and function. The comment block would need to also need to be located where DartDoc can process it.

The documentationcheck.py file could then be updated to flag all files that don't comply. It currently just checks modified files in the PR to see whether there is at least one comment line. It does nothing more. This is insufficient.

With this approach we can slowly get the documentation improved.

This is the minimum requirement.

Still interested?

**Comment by literalEval** [comment]
I do know python and have written quite some scripts for daily use. 

So is it like that the script will run on each file, parse it and check whether the
- Comments are present
- At the right place
- In the right format 

with string parsing techniques ?

**Comment by palisadoes** [comment]
Yes.

It would need to report:
1. The filename
1. The line number where the error was found
1. The method/class/function where it was found
1. There argument(s)/return value(s) that are not documented where applicable
1. The general description that may be missing

It would also need to:
1. do this whether presented with filename(s) or a directory(ies) as an argument
1. recursively check files in the directory

User python's argparse module to differentiate between filenames or the directories.

The script would need to show these results after processing all the data presented, not just after the first file is analyzed.

Also, ensure you:

1. Follow the Google python style guide with your code.
1. Format the code using python black
1. Run the pylint linter on the code

Black and pylint could also be added to our GitHub actions to improve the quality of our python helper scripts.

The next issue to be created would be to do the same Google style analysis on our python scripts as an additional GitHub action. This would be useful for all our repositories.

Please take the time to do a good job. This is an important thing to fix.

**Comment by literalEval** [comment]
For now I am looking at custom linting in Flutter and will try to make one from scratch if I feel it can work.
If custom linting doesn't work, I will look at the Python script option and try file parsing.

**Comment by literalEval** [comment]
@palisadoes @anwersayeed I wanted to ask what is our target for documentation ? Is it enough if public classes and their 
 methods, attributes etc are documented ?

**Comment by literalEval** [comment]
@palisadoes @anwersayeed I have found [this](https://dart-lang.github.io/linter/lints/public_member_api_docs.html) lint option that asks for documentation for 'all public classes and their members'. Will it do ?

**Comment by palisadoes** [comment]
> @palisadoes @anwersayeed I have found [this](https://dart-lang.github.io/linter/lints/public_member_api_docs.html) lint option that asks for documentation for 'all public classes and their members'. Will it do ?

1. This is a good first step. Can the linter be applied to a single file so that it can be used by our existing PR python script?
2. We still need the check for standardized comments in classes, functions and methods.

**Comment by literalEval** [comment]
@palisadoes 

1. I will check and update.
2. I already works for all classes and their public members (both attributes and methods) along with non class things like normal functions, enums etc. The only limitation being that it only work for things which are public and won't check for documentation of private fields. Is it okay ?

**Comment by palisadoes** [comment]
1. @literalEval That should be OK. Let's see what you have. What language are you using to do the linting?
1. @noman2002 @TheHazeEffect You should review this issue and PR. It should greatly improve our code quality and documentation

**Comment by literalEval** [comment]
@palisadoes what I am saying is that there already exists a lint rule in Flutter which when turned on, will

1. Show warning for every public class and its methods/attributes which are not documented
2. Can be run from command line for any specific file and show an error if it doesn't find documentation.

I am asking is will this much of documentation enforcement do ? Or we need to strictly check for documentation of private fields too ?

**Comment by palisadoes** [comment]
What you showed only checks for the existence comments in the right location. It doesn't check for the content of the comments. 
1. We can implement your suggestion as it's better than what we have now.
2. We still need the basis of this issue to be implemented.

**Comment by literalEval** [comment]
@palisadoes  Hmm. I have thought of two possibilites.

1. Write a Python script which will check for existence of documentation and its format. I have started writing this script and have some success in basic checks. Basically I am doing this by simple string parsing.
2. Use [this custom lint plugin](https://pub.dev/packages/custom_lint) and implement this functionality as native dart feature which can even be integrated to IDEs (as it directly works with dart analyzer).

I am currently working on the 1st option and will let you know about the progress.

**Comment by palisadoes** [comment]
@literalEval Please remember the aim is to enforce compliance as a GitHub action.

**Comment by literalEval** [comment]
Hmm. I'll try to make a basic working prototype and update in 2-3 days. This issue will obviously take long to complete :)

**Comment by literalEval** [comment]
@palisadoes I think I have got what we needed.

- After careful consideration, I thought that writing a python script will be a quite bad idea. As either I will write a simple parser, which by design, won't be extensible and will take a lot of effort to be fail safe, or I'll end up writing a full blown parser from scratch, which makes no sense as there is already a linter available for Dart.
- The best solution (of course) will be to write an Abstract Syntax Tree (AST) parser, but that will take too much work and neither can a single person do that for a proper language.

Looking at the problem again - 

1. Since [a lint option that checks for documentation of public classes, methods etc](https://dart-lang.github.io/linter/lints/public_member_api_docs.html) already exists and we only need to extend it to private fields as well.
2. Since this is just a dart script, it can be run from command line.
3. If we are sure that documentation for every fields exists, a simple string parser script can be written in Python that checks if the given documentation is in right format or not. I think this is all we need. Isn't it ?

So I came up with a solution for problem 1 - 

1. Since `public_member_api_docs` checked for public fields, it must be doing some checks to find if a field is public or not. I forked up the [default dart linter](https://github.com/dart-lang/linter) and modified the `public_member_api_docs` rule. After removing the public field checks, the linter was giving errors for ALL fields if they are not documented.

For `lib/views/main_screen.dart`, which is like

![image](https://user-images.githubusercontent.com/62198564/218725105-95a7f4ed-aeac-43b1-acf3-7ec8b6d9cc43.png)

this modified linter gives below errors.

![image](https://user-images.githubusercontent.com/62198564/218724171-01ffe902-72fa-49e1-87a9-eb9b714ffbfe.png)

Now it checks for documentation for ALL the fields. The only problem that remains is that it will also ask for documentation of class constructors etc. but I am sure I can further modify it to not do so.

Is this good progress and should I continue in the same direction ? As I think there can't be any better idea than using the official linter itself. Please share your review and also any suggestions if you have.

**Comment by palisadoes** [comment]
1. What does it need to pass? What format does the documentation need to have?
1. If you get it fixed we need to add it as an option for the dart linter and submit a PR in the necessary repository. This will mean it will only be available for the latest versions of flutter. Do you see that as an issue?
1. With your solution, will we need to have our own linting repo until the official Dart one or can we place it on our `.github` directory? Would it be one file or a set of libraries?

It's good to see progress on this

**Comment by palisadoes** [comment]
@TheHazeEffect , @Sagar2366, @noman2002 what do you think about this issue and the protected solution?

**Comment by literalEval** [comment]
1. It needs the path to file or directory. I will write the checks so that it reports compliance with DartDoc guidelines.
2. We don't need to add this into lint options. We can run this as separate script just like we are doing with python scripts.
3. Yes. We will store the fork in our workflow. After removing unnecessary files from it which we don't need, it would be of around 1-2 mb I guess.

**Comment by palisadoes** [comment]
This is not practical. Including a complete linter in the code base will add bloat that will have to maintain.

Surely there must be a simple way to:
1. Read a file
1. Find the method class and functions
1. Determine whether there are comments above these locations
1. Determine whether the very first comment has text that doesn't refer to the arguments
1. Determine whether the comments mention the arguments in a specified format that a regular expression could do.
1. Make sure there are lines between the comments for readability.


This should not be hard. You are not giving this enough thought.

**Comment by literalEval** [comment]
And how about I create a plugin (this modified linter) in my GitHub and for talawa we only refer to that plugin instead of adding the whole thing to the codebase ? 
Also, creating custom lints isn't a new thing for orgs. While finding solutions to this problem, I found out that many orgs maintain their own custom lints.

Sure it can be done using string parsing with some python script but using a real linter is the most fail safe solution. But still, if everyone finds this linter idea bad, I'll try to write a python script.

**Comment by palisadoes** [comment]
I respect the enthusiasm, but some of your solutions only fix the short term, but not long term issue.
1. If we use your repo, will we be able to rely on you over the years to do updates?
1. `custom_lint` would be a better solution. Will you be able to implement the logic required for this issue in Dart?
     1. For the entire repo
     1. For a list of files.     
1. Will `custom_lint` require us to have to maintain multiple MB of supporting code? For a small community like ours, we'll need something that is easily manageable. A few hundred lines of code for now.

**Comment by palisadoes** [comment]
Thanks. 
1. I'm not in favor of this depending on a singe contributor repo. Life circumstances change and it's always best to rely on multi-person repos. I have to think of the all the Talawa repo stakeholders.
1. Please proceed only with the `custom_lint` and the `public_member_api_docs` options where applicable.

**Comment by literalEval** [comment]
Sure. I'll try my best to make the custom_lint integration as smooth as possible. 
I'll try to update soon.

**Comment by literalEval** [comment]
@palisadoes I have successfully integrated `custom_lint` in 'talawa'. Adds up very well and even gives lint warnings/errors in the IDE itself. Changes that will follow post integration - 

1. Another folder will be added in the talawa root folder (just like we have one for tests) and will contain all the lint rules which we want to enforce. Although we can put everything in a single file, but still maintaining code cleanliness, 2 files for our 2 proposed rules will be added. So an addition of 1 directory and 2-3 files.
2. 3 dev dependencies will be added. Since they are dev dependencies, nothing will change for the user and apk size.
3. `custom_lint` can be run from the command line in our GitHub workflow and will fail if any file is found to not follow our lint rules.

Only issue for now is that it can only scan an entire package and not a single file or list of files. For this, I am actively discussing with the `custom_lint` plugin maintainer to add the feature [here](https://github.com/invertase/dart_custom_lint/issues/80). I have even made it work locally and if he approves, I will push the changes to their repo.

Your views ?

**Comment by palisadoes** [comment]
Thanks for the update. To complete this issue you'll need to prove that:

1. It can work recursively through a directory on all files found
1. That it can work on a list of files from 1 - N 
1. That it can be integrated into GitHub actions

Let us know when the `custom_lint` PR has been merged

**Comment by literalEval** [comment]
Will do. Thanks.

**Comment by literalEval** [comment]
@palisadoes I opened the respective PR in `custom_lint` repo [here](https://github.com/invertase/dart_custom_lint/pull/82). The maintainer seems to be busy maybe. 
Since I have already implemented the 2 custom lint rules as discussed in this issue, how about we put all the files in the exclude list of `custom_lint` so that no file gets scanned initially and then we remove files from the exclude list with each new PR ? Meanwhile if my PR in `custom_lint` gets merged and they release a new version with changes, then we will do the checking properly.

Do you like this idea ? Or should I just wait for the maintainer to merge my PR ?

**Comment by palisadoes** [comment]
Is there a way in which we could dynamically remove files from the ignore list during the PR and apply the custom linter rules?

Manual changes are always difficult to make consistent

**Comment by literalEval** [comment]
@palisadoes sure. I have thought about the whole procedure.

1. Scan all of the codebase and add `// ignore: our_custom_lint_rule` to the top of all of our files. A simple python script will do this.
2. New PR comes and according to contribution rules (which we will update), contributor should remove this lint suppression directive and add documentation for changed files.
3.  We scan for the changed files with a Python script and check if they contain `// ignore: our_custom_lint_rule` anywhere (not only on the top because then someone could insert that line anywhere in between the file to bypass the lint analysis). If found, the scan fails, otherwise passes. 
4. If the scan is passed, we can now safely run our custom lint rules on the codebase. 

Your views @palisadoes ?

**Comment by palisadoes** [comment]
This strategy would also mean that any new files would automatically fail because they would not have the string.
Your script will need to work for both pull-requests and pushes. That's in the spirit of the change.

For example, the documentationcheck.py currently works for pushes because the GitHub action `github` parameter that it uses checks for the branch name, but in PRs the value presented is different.

- [Sample output for PR](https://github.com/PalisadoesFoundation/talawa/actions/runs/4190764070/jobs/7264494864#step:10:1013)
- [Sample output for Push](https://github.com/PalisadoesFoundation/talawa/actions/runs/4190623837/jobs/7264218538)

As part of the PR you will therefore need to:
1. test your script in both cases.
2. prove that the linter complies with the requirement

**Comment by literalEval** [comment]
I think I wasn't able to convey my point clearly.

1. The said script will fail **if that string IS present**.
2. New files won't have that `// ignore` directive and would automatically pass from the script and will be checked for documentation.

About the 2nd point, sure, I will look into these parameters thoroughly and make the script accordingly.

**Comment by palisadoes** [comment]
How can you design it make it work for new files too?

**Comment by literalEval** [comment]
Sorry, I am unable to understand your query @palisadoes. Do you mean how will this script + custom_lint work together on any new file ? Or how would I get the name of a new file in Github action ?

**Comment by palisadoes** [comment]
Your proposed solution only works for files currently in the repo. How do we get it to work in an automated way when new files are added?

**Comment by literalEval** [comment]
Let me try to explain again - 

1. We add a `// ignore: ....` directive to all the existing files, so that none of them are scanned initially.
2. Then with each new PR, we check which files are modified are created as new. 
3. We scan these new/modified files for the `ignore directive`. If they contain such a line anywhere, the test fails, otherwise passes.
4. We run our `custom lint analyzer` on the codebase. Since we are sure that the modified/new files don't have that ignore directive, that mean the analyzer will surely analyze them for documentation.

Does this make sense ?

**Comment by palisadoes** [comment]
1. Your script will ensure that now files submitted in the pull request have the ignore statement.
1. If that passes, then the linter runs against the entire code base in which case only the submitted files are scanned.

Correct?

**Comment by literalEval** [comment]
1. Yeah. It will ensure both new as well as modified files.
2. Yeah just like `flutter analyze` in our github actions runs on entire codebase, our `custom_lint_rule` will run too.

I think I was able to make you understand the idea clearly.

**Comment by palisadoes** [comment]
OK please go ahead

**Comment by literalEval** [comment]
@palisadoes I wanted to ask that sometimes we don't really need to document an argument as is might be really self descriptive. So do we really need to enforce documentation of parameters etc ? Or they can be left to be on voluntary basis ?

**Comment by literalEval** [comment]
Okay. Will update till weekend, since exams are going on.

**Comment by palisadoes** [comment]
Also lint for print statements: https://github.com/PalisadoesFoundation/talawa/issues/1550#issuecomment-1438697447

**Comment by literalEval** [comment]
@palisadoes progress. Please review.

## Functions/Methods

- No doc

![Screenshot_20230224_235520](https://user-images.githubusercontent.com/62198564/221260102-8d19bad1-57a9-49ae-9773-4fcef5fccd54.png)

- First line doesn't end with 'end punctuation'.

![Screenshot_20230224_235701](https://user-images.githubusercontent.com/62198564/221260438-4fc90b28-844c-4778-99e3-151b5efa8877.png)

- Second line is not empty. (Second error is about not containing param)

![Screenshot_20230224_235855](https://user-images.githubusercontent.com/62198564/221261025-be59041d-7cae-43b6-95c7-fc6e5b9005fe.png)

- '*' should follow with corresponding **param** name

![Screenshot_20230225_000146](https://user-images.githubusercontent.com/62198564/221262069-9e89d946-1f2d-49e5-96c7-0e9006a11888.png)

- Not all parameters are documented

![Screenshot_20230225_000359](https://user-images.githubusercontent.com/62198564/221262653-48a74541-bd96-46e6-9953-7694d3441653.png)

- If all the rules are followed

![image](https://user-images.githubusercontent.com/62198564/221262800-0fb76957-4118-4755-a9d8-ed4a5c52fc26.png)

## Others

Very similar, the only difference being that we don't need parameters in them.

## Command Line

- Fails. When 'non-ignored' files have lint issues. It gives a `error` return code.

![image](https://user-images.githubusercontent.com/62198564/221268101-f167eb08-a01c-4a41-b287-ae7a4fdf9bc0.png)

- Passes. When 'non-ignored' files have no lint issues.

![image](https://user-images.githubusercontent.com/62198564/221268442-30de8cc7-636e-4929-8b0c-f8ba49e7e08e.png)

**Comment by literalEval** [comment]
The actual errors generated can be further customized and the error behavior too.

**Comment by palisadoes** [comment]
Please:
1. check that there is some text after the colons
1. check that there is a `returns:` block to describe what the expected results should be. `None` should be acceptable if nothing is returned.

**Comment by palisadoes** [comment]
Are the screenshots from Android Studio?

**Comment by literalEval** [comment]
1. Yeah that is done. Forgot to show.
2. Noted.
3. No. It is VSCode. You want me to check on Android Studio too ?

Thanks for the review @palisadoes

**Comment by palisadoes** [comment]
Yes, we should verify Android Studio.

**Comment by literalEval** [comment]
@palisadoes works on Android Studio. No extra setup is required.

**Comment by literalEval** [comment]
> Please:
> 
> 1. check that there is some text after the colons
> 2. check that there is a `returns:` block to describe what the expected results should be. `None` should be acceptable if nothing is returned.

@palisadoes shouldn't the `returns:` and `params:` be optional if the function returns nothing and has no parameters respectively ? Because the IDE itself will show that the function has `void` return type. We don't need documentation for that case I think. Same with empty parameters.

**Comment by literalEval** [comment]
@palisadoes updates

- Nothing follows the params

![Screenshot_20230225_165203](https://user-images.githubusercontent.com/62198564/221354211-e5398aaa-77d5-4d38-9d31-40db4c775123.png)

- `returns:` is not needed for `void` type, but if present, it should be followed by `None`

![Screenshot_20230225_170050](https://user-images.githubusercontent.com/62198564/221354562-ec6931b8-5f12-460b-9781-242909e33c83.png)

![Screenshot_20230225_170212](https://user-images.githubusercontent.com/62198564/221354610-604c1d35-3ccf-48be-93d7-161bf2d538c3.png)

![Screenshot_20230225_170257](https://user-images.githubusercontent.com/62198564/221354653-5daf8f98-59e1-4558-8797-3ce0373ad537.png)

- `returns:` must be present for non-void types and should be followed by documentation.

![Screenshot_20230225_170506](https://user-images.githubusercontent.com/62198564/221354751-39817a54-5a69-4061-aa62-6ede141a44e3.png)

![Screenshot_20230225_170622](https://user-images.githubusercontent.com/62198564/221354779-88ef61aa-fb89-4c6c-800c-a170f5863b9a.png)

**Comment by literalEval** [comment]
@palisadoes what do you think about the progress ? :)

**Comment by palisadoes** [comment]
What I meant was:

```
returns:
     None
```
or
```
returns:
     result: Integer value representing etc..... 
```

Just like we have with the arguments

**Comment by palisadoes** [comment]
That should be a quick fix. I really want to get this in place.

**Comment by literalEval** [comment]
1. About the `void` return, isn't the single line syntax better ? As it takes less space. Otherwise we'll be wasting another line just to say `None`
2. Understood. I believe the `result` should be the name of the variable that is returned ? Right ? Something like this ?
```
/// returns:
///     `mAge:` Age of the owner.
int myAge() {
    int mAge = 10;
    return mAge;
}
```

Also, have you thought about the `returns:` being optional when the return type is `void` ?

**Comment by literalEval** [comment]
The reason I am being so lenient on the return thing is because the Description documentation of the function/method itself will tell about the stuff the function does.

```
/// Returns the age of the owner.
/// 
/// returns:
///     `mAge:` Age of the owner.
int myAge() {
    int mAge = 10;
    return mAge;
}
```

Something like this. Isn't the `returns:` explanation being redundant here ?

**Comment by literalEval** [comment]
@palisadoes @noman2002 what do you think ?

**Comment by palisadoes** [comment]
For tiny functions, yes. But most functions are more complex. 

Remember, the reason for having this requirement is that we will be automatically extracting this data with DartDoc and adding it the Talawa website so that people can understand what all code does without having to examine it. In other words it improve our reference materials.

**Comment by literalEval** [comment]
That I totally understand @palisadoes. I have other ideas but maybe we can discuss them when this is merged, as version-2 of this implementation :)

**Comment by literalEval** [comment]
Can the void return doc be like `returns: None` or having a separate line is a must ?

**Comment by palisadoes** [comment]
Yes, I was thinking python which uses "none"

**Comment by literalEval** [comment]
Sorry @palisadoes  I asked two questions at once and can't see which one you agreed on :) 
So a separate line is a must and `returns: None` in a single line shouldn't be acceptable. Right ?

**Comment by palisadoes** [comment]
Yes. 
- Returns void, with the same expected layout that the arguments have with the heading on a different line from the variable descriptions

**Comment by literalEval** [comment]
> 1. About the `void` return, isn't the single line syntax better ? As it takes less space. Otherwise we'll be wasting another line just to say `None`
> 2. Understood. I believe the `result` should be the name of the variable that is returned ? Right ? Something like this ?
> 
> ```
> /// returns:
> ///     `mAge:` Age of the owner.
> int myAge() {
>     int mAge = 10;
>     return mAge;
> }
> ```
> 
> Also, have you thought about the `returns:` being optional when the return type is `void` ?

Like the 2nd point ?

**Comment by literalEval** [comment]
OR

it should be

```
/// returns:
///     `int`: age of owner
```

?

**Comment by palisadoes** [comment]
Let's leave that decision up to the coder. As long as they do the documentation we'll be much better off than we are today.

**Comment by literalEval** [comment]
Okay. I'm going with the 2nd option for now.

**Comment by evasharma12** [comment]
Hey @literalEval, I want to confirm that the HTML code will be generated smoothly with the linter.
1. Install [Dart Doc](https://pub.dev/packages/dartdoc) and then run the command `dart doc .`
2. Modify some file's documentation according to the new format and then generate the HTML docs using the above command.
3. Post the screenshot for the same.

**Comment by tasneemkoushar** [comment]
> I think I wasn't able to convey my point clearly.
> 
> 1. The said script will fail **if that string IS present**, in any of the changed files.
> 2. New files won't have that `// ignore` directive and would automatically pass from the script and will be checked for documentation.
> 
> About the 2nd point, sure, I will look into these parameters thoroughly and make the script accordingly.

@literalEval 
1. As I can see that the PR is merged, Is your documentation linter for dart doc comments working for new PRs now? As I tried to test it by altering comments and making the syntax wrong, but no error was reported!

2. When will you update Contributing.md for new PRs?

3. Won't it be confusing for people to remove the flint suppression directive when making a pull request?

**Comment by literalEval** [comment]
@tasneemkoushar 

1. Final integration is not done yet. The command that checks for documentation in the codebase is present there but is commented out. It will be uncommented as the final step. For now, two steps remain
    - Add a script that checks if ignore directive is present or not. If present, it will fail, otherwise pass. I have made this script and the only integration with workflow remains.
    - Uncomment the lines from workflow which check for custom lint rules.
2. In the final step.
3. I don't think so. `Contributing.md` will explicitly mention this point and the workflow errors will too. If you think otherwise, please elaborate any alternative. Also, as you can see our discussion (with @palisadoes ) a couple of comments ago, you will see that I was also working on adding a feature to `custom_lint` that will allow us to check individual/group of files. That PR is pending and I will also try to get that merged. This ignore suppression is just the alternative to that.

Hope I answered all of your queries. Thanks.

**Comment by literalEval** [comment]
@tasneemkoushar can you answer my queries on Slack ? You are the possible mentor of the idea I wanted to work on for GSoC period. Thanks.

**Comment by tasneemkoushar** [comment]
> @tasneemkoushar can you answer my queries on Slack ? You are the possible mentor of the idea I wanted to work on for GSoC period. Thanks.

Sure, Write me with all your queries on slack, I will get back to you as soon as I get time.
